### ISSUE TITLE
Add full text search to GraphQL and TerminusDB

### ISSUE THREAD

**Issue by GavinMendelGleason** [terminusdb/terminusdb#1516]

**Comment by hoijnet** [comment]
This would be awesome, very curious how the query interface will look and a potential relationship with WOQL for full text search parametrisation!

Looking forward to follow!

**Comment by GavinMendelGleason** [comment]
I'll try to provide more information here soon. We are definitely taking inspiration from DGraph here.

**Comment by dfrnt-SigmundFuchs** [comment]
This issue has been closed due to inactivity. Please don’t hesitate to reopen it if you have any further questions or would like to continue the discussion.

**Comment by MentalGear** [comment]
I'd be interested if there are new developments for text or vector search, as discussed in https://github.com/orgs/terminusdb/discussions/1432#discussioncomment-14823744 .

**Comment by MentalGear** [comment]
Thank you for the reply! Good to see there is increasing interest in around this feature.

I'd like to hear about the historical searchability in FTS context, is it somewhere documented already?

The use-case is (of course :P) discoverability for llm-agents.

**Comment by hoijnet** [comment]
It is not condensed into stable form yet. The idea is in early stages, with some ideas internally about architecture for it. There are many aspects to consider and it's important to get it right!

**Comment by hoijnet** [comment]
There are some developments happening here on both the vector search and full text search front where we are taking the ideas behind vectorlink further. The exact shape is yet to be determined but a first step is the ability to surface embeddings as a document "format" so that you can see what the embeddings look like rendered which lands in 12.0.6; and more is on the roadmap.

Reach out directly on discord if interested to testing it early!

**Comment by hoijnet** [comment]
> I'd like to hear about the historical searchability in FTS context, is it somewhere documented already?

Indeed. Inspecting the automatic embeddings was just shipped in version 12.0.6. A proper search solution should support both FTS, vector search and hybrid modes I think. Also, a solution like TerminusDB should have full historic search without indexing every version of the data.

There's work ongoing in this area. Timeline still a bit unclear. How are things going on your end?

**Comment by MentalGear** [comment]
Thank you for the updates !

I really appreciate the efforts on these essential search features! Having agents being to search the full branching of a decision tree is crucial in finding optimal paths.

For some prior work on balancing full-text, vector, and hybrid search, this engine might be interesting: https://docs.orama.com/docs/orama-js/search/hybrid-search#custom-hybrid-weights

Looking forward to what you come up & also well done on the new docs pages !

PS: It would be great to have other options available other than openAI for generating vectors. There are many plug-and-play adapters that abstract the LLM provider, and having the option to use something EU sovereign like Mistral or just specific  (e.g. Anthropic) can be a project requirement as the embeddings might be more fitting.

**Comment by hoijnet** [comment]
Thanks, looking forward to your feedback once there!

> It would be great to have other options available other than openAI for generating vectors. There are many plug-and-play adapters that abstract the LLM provider, and having the option to use something EU sovereign like Mistral or just specific (e.g. Anthropic) can be a project requirement as the embeddings might be more fitting.

I fully agree, I believe what most will want is the self-sovereign option as well for embeddings; as well as being able to use for example open router and any OpenAI endpoint in general. Thanks for confirming that direction is valid!

All input is welcome for improving the docs pages as well; tried to make it as easy as possible to suggest improvements as well!

**Comment by hoijnet** [comment]
For anyone wanting to play with vectorlink, I [revived the old vectorlink repo](https://github.com/terminusdb-org/vectorlink) and added support for running with a local ollama endpoint with for example quen3:4b embeddings.

This is just to ensure it's still supported, tested and backwards compatible in the upcoming 12.1 release as a first step. The support will move into an optional dedicated plugin for integrating with vectorlink. If you use it, give a shout!

The integration tests and the start script should help you get started. That said, keep your eyes open for more in the vector and hybrid search space.

At least it's a start to know this is still workable.

**Comment by hoijnet** [comment]
This is now part of the upcoming 12.1-rc release, with both vector search and full text search. I'll post notes on how to get started as part of a information related to 12.1-rc.

We are using ollama and nomic v2 in the initial version, more embeddings models to come. What are the suggested open and closed embedding models that you think should be supported?

**Comment by thesmartwon** [comment]
> What are the suggested open and closed embedding models that you think should be supported?

Colbert-v2 models like [mixedbread's](https://huggingface.co/collections/mixedbread-ai/mxbai-edge-colbert-v0-series) which require late interaction. Perhaps a [GEM index](https://arxiv.org/pdf/2603.20336) could work?

**Comment by hoijnet** [comment]
Thanks for the input! What makes Colbert-v2 a strong contender? It does not show up as popular alternatives generally do became quite curious! I'm unsure if the core would support it with our use of LanceDB would work well with it? Any experiences?

The research since the question has pivoted my thinking from nomic-embed-v2 (which very few are hosting) to bge-v3 which seems widely supported and quite effective (versatile) for the general population and make it the interoperable default in TerminusDB unless the user ejects and picks their own.

Getting it right enough is helpful.

If Colbert-v2 is indeed the better option, it would be great to understand the reasoning behind.

**Comment by thesmartwon** [comment]
> What makes Colbert-v2 a strong contender?

Better precision and recall than dense models and a late document/query interaction matrix that can be used for snippets and highlighting.

> It does not show up as popular alternatives generally do

That's true, according to huggingface the most popular model only has [2M downloads](https://huggingface.co/models?other=ColBERT&sort=downloads) to another embedder's [65M](https://huggingface.co/models?pipeline_tag=feature-extraction&sort=downloads). It's unpopular because of its high friction.

> I'm unsure if the core would support it with our use of LanceDB would work well with it?

It wouldn't without using one compatible with a [MUVERA](https://arxiv.org/abs/2405.19504) shim. Without MUVERA, you need a special index like [PLAID](https://lightonai.github.io/pylate/documentation/retrieval/) or the more recent [GEM](https://arxiv.org/pdf/2603.20336) for large corpuses.

> Any experiences?

These models are most often used as rerankers, although I personally like using them instead of dense models altogether because of the interaction matrix insight. With dense models you're left guessing what the dimensions mean.

> The research since the question has pivoted my thinking from nomic-embed-v2...

Whatever model you choose today as the default will be likely no longer be Pareto optimal (in size/performance) in less than a year. That's what's happened to nomic-embed-v2 and bge-v3 already.

> Getting it right enough is helpful.

"Right" is different based on application. Multilinguality should be a flag because the models tend to have twice the inference cost for the same performance as an English-only model. Size should be tuned to whatever default GPU you use.

You can check the [MTEB leaderboards](https://huggingface.co/spaces/mteb/leaderboard) and filter by various constraints to pick reasonable defaults.

For multilingual, [Jina AI](https://huggingface.co/jinaai) has been in the lead for a while now at smaller weight sizes and Gemma or Qwen are best at larger sizes.
